File: src/deserialize/reader.c
#include <stdio.h>
#include <utils.h>
#include "con_reader.h"

size_t con_reader_comment_read(void const *context, char *buffer, size_t buffer_size);

enum ConError con_reader_comment_init(struct ConReaderComment *context, struct GciInterfaceReader reader) {
    if (context == NULL) { return CON_ERROR_NULL; }
    context->reader = reader;
    context->state = con_utils_state_char_init();
    context->buffer_char = EOF;
    context->in_comment = false;
    return CON_ERROR_OK;
}

struct GciInterfaceReader con_reader_comment_interface(struct ConReaderComment *context) {
    return (struct GciInterfaceReader) { .context = context, .read = con_reader_comment_read };
}

size_t con_reader_comment_comment_start(struct ConReaderComment *context, char *buffer, size_t buffer_size) {
    assert(context != NULL);
    size_t length = 0;

    char c;
    size_t l = gci_reader_read(context->reader, &c, 1);
    assert(l == 0 || l == 1);

    if (l != 1) {
        length = 0;
    } else if (c == '/') {
        length = 0;
        context->buffer_char = EOF;
        context->in_comment = true;
    } else {
        context->buffer_char = EOF;
        buffer[length++] = '/';

        if (length >= buffer_size) {
            context->buffer_char = c;
        } else {
            buffer[length++] = c;
        }
    }

    return length;
}

size_t con_reader_comment_read(void const *void_context, char *buffer, size_t buffer_size) {
    assert(void_context != NULL);
    struct ConReaderComment *context = (struct ConReaderComment*) void_context;

    bool any_read = false;
    size_t length = 0;

    assert(buffer != NULL);
    if (context->buffer_char != EOF) {
        if (context->buffer_char == '/') {
            assert(!context->in_comment);
            size_t l = con_reader_comment_comment_start(context, buffer, buffer_size);
            if (l == 0 && !context->in_comment) { return 0; }
            length += l;
        } else if (buffer_size >= 1) {
            buffer[0] = (char) context->buffer_char;
            context->buffer_char = EOF;

            length = 1;
            any_read = true;
        } else {
            return 0;
        }
    }

    while (length < buffer_size) {
        char c;

        size_t l = gci_reader_read(context->reader, &c, 1);
        assert(l == 0 || l == 1);
        if (l != 1) { break; }

        if (!context->in_comment && !con_utils_state_char_is_string(context->state) && c == '/') {
            l = con_reader_comment_comment_start(context, buffer + length, buffer_size - length);
            if (l == 0 && !context->in_comment) {
                context->buffer_char = '/';
                break;
            }

            length += l;
        } else if (context->in_comment && c == '\n') {
            context->in_comment = false;
            buffer[length++] = '\n';
        } else if (!context->in_comment) {
            buffer[length++] = c;
        }

        con_utils_state_char_next(&context->state, c);
        any_read = true;
    }

    assert(length <= buffer_size);
    return length;
}

File: src/deserialize/reader.zig
const gci = @import("gci");
const internal = @import("../internal.zig");
const lib = internal.lib;

pub const Comment = struct {
    inner: lib.ConReaderComment,

    pub fn init(reader: gci.InterfaceReader) !Comment {
        var self: Comment = undefined;
        const err = lib.con_reader_comment_init(
            &self.inner,
            @as(*lib.GciInterfaceReader, @ptrCast(@constCast(&reader.reader))).*,
        );
        try internal.enumToError(err);
        return self;
    }

    pub fn interface(self: *Comment) gci.InterfaceReader {
        const temp: gci.InterfaceReader = undefined;
        return .{ .reader = @as(
            *@TypeOf(temp.reader),
            @ptrCast(@constCast(&lib.con_reader_comment_interface(&self.inner))),
        ).* };
    }
};

const testing = @import("std").testing;

test "comment init" {
    const d = "";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    _ = context.interface();
}

test "comment read" {
    const d = "12";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    const reader = context.interface();

    var buffer: [2]u8 = undefined;
    const result = try reader.read(&buffer);
    try testing.expectEqualStrings("12", result);
}

test "comment read comment" {
    const d = "[  //:(\n \"k //:)\",1/]";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    const reader = context.interface();

    var buffer: [17]u8 = undefined;
    const result = try reader.read(&buffer);
    try testing.expectEqualStrings("[  \n \"k //:)\",1/]", result);
}

test "comment read comment one char at a time" {
    const d = "[  //:(\n \"k //:)\",1/]";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    const reader = context.interface();

    var buffer: [17]u8 = undefined;
    for (0..17) |i| {
        const result = try reader.read(buffer[i .. i + 1]);
        try testing.expectEqual(1, result.len);
    }

    try testing.expectEqualStrings("[  \n \"k //:)\",1/]", &buffer);
}

test "comment inner reader empty" {
    const d = "";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    const reader = context.interface();

    var buffer: [1]u8 = undefined;
    const err = reader.read(&buffer);
    try testing.expectError(error.Reader, err);
}

test "comment inner reader empty comment" {
    const d = "/";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    const reader = context.interface();

    var buffer: [1]u8 = undefined;
    const err = reader.read(&buffer);
    try testing.expectError(error.Reader, err);
}

test "comment inner reader fail" {
    var c1 = try gci.ReaderString.init("1");
    var c2 = try gci.ReaderFail.init(c1.interface(), 0);

    var context = try Comment.init(c2.interface());
    const reader = context.interface();

    var buffer: [1]u8 = undefined;
    const err = reader.read(&buffer);
    try testing.expectError(error.Reader, err);
}

test "comment inner reader fail comment" {
    var c1 = try gci.ReaderString.init("/");
    var c2 = try gci.ReaderFail.init(c1.interface(), 1);

    var context = try Comment.init(c2.interface());
    const reader = context.interface();

    var buffer: [1]u8 = undefined;
    const err1 = reader.read(&buffer);
    try testing.expectError(error.Reader, err1);

    const err2 = reader.read(&buffer);
    try testing.expectError(error.Reader, err2);
}

test "comment read only comment" {
    const d = "// only a comment";
    var c = try gci.ReaderString.init(d);

    var context = try Comment.init(c.interface());
    const reader = context.interface();

    var buffer: [3]u8 = undefined;
    const err1 = reader.read(&buffer);
    try testing.expectError(error.Reader, err1);
}

test "comment read clear error" {
    var c1 = try gci.ReaderString.init("1");
    var c2 = try gci.ReaderFail.init(c1.interface(), 0);

    var context = try Comment.init(c2.interface());
    const reader = context.interface();

    var buffer: [2]u8 = undefined;
    const err = reader.read(&buffer);
    try testing.expectError(error.Reader, err);

    c2.inner.reads_before_fail = 1;
    const result = try reader.read(&buffer);
    try testing.expectEqualStrings("1", result);
}

test "comment read comment clear error" {
    var c1 = try gci.ReaderString.init("//\n1");
    var c2 = try gci.ReaderFail.init(c1.interface(), 1);

    var context = try Comment.init(c2.interface());
    const reader = context.interface();

    var buffer: [3]u8 = undefined;
    const err = reader.read(&buffer);
    try testing.expectError(error.Reader, err);
    try testing.expectEqual('/', context.inner.buffer_char);

    c2.inner.reads_before_fail = 4;
    const result = try reader.read(&buffer);
    try testing.expectEqualStrings("\n1", result);
}

test "comment reader half comment clear error" {
    var c1 = try gci.ReaderString.init("/1");
    var c2 = try gci.ReaderFail.init(c1.interface(), 1);

    var context = try Comment.init(c2.interface());
    const reader = context.interface();

    var buffer: [3]u8 = undefined;
    const err = reader.read(&buffer);
    try testing.expectError(error.Reader, err);
    try testing.expectEqual('/', context.inner.buffer_char);

    c2.inner.reads_before_fail = 2;
    const result = try reader.read(&buffer);
    try testing.expectEqualStrings("/1", result);
}
